components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/util/ClientCredentialAuthConfig.java

```diff
@@ -0,0 +1,392 @@
+                authenticationContext.getTenantDomain());
+        try {
+            if (StringUtils.isNotEmpty(accessToken) && !isTokenExpired(accessToken)) {
+                LOG.debug("Unexpired access token available in cache.");
+                return accessToken;
+            } else {
+                // Attempt the first request for an access token
+                LOG.info("Attempting initial access token request for session data key: " +
+                        authenticationContext.getContextIdentifier());
+                accessToken = requestAccessToken(); // Attempt to request a new token
+                if (accessToken != null) {
+                    return accessToken;
+                }
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_TOKEN);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.TOKEN_ENDPOINT, getTokenEndpoint())
+                            .configParam(Constants.LogConstants.ConfigKeys.SUPPORTED_GRANT_TYPES,
+                                    GRANT_TYPE_CLIENT_CREDENTIALS)
+                            .resultMessage("Initial request failed, proceeding with retry attempts.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                // If the initial request fails, proceed with retry logic
+                LOG.info("Initial request failed, proceeding with retry attempts.");
+                return attemptAccessTokenRequest(maxRequestAttemptsForAPIEndpointTimeout);
+            }
+        } catch (ParseException e) {
+            LOG.error("Error parsing token expiry.", e);
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+        } catch (IllegalArgumentException e) {
+            LOG.error("Invalid endpoint URL: " + getTokenEndpoint(), e);
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+        } catch (Exception e) {
+            LOG.error("Unexpected error during token acquisition.", e);
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+        }
+        return null;
+    }
+
+    /**
+     * This method is used to attempt the access token request from the token endpoint.
+     *
+     * @param maxAttempts Maximum number of attempts to request the access token
+     * @return Access token
+     */
+    private String attemptAccessTokenRequest(int maxAttempts) {
+
+        int attemptCount = 0;
+
+        while (++attemptCount < maxAttempts) {
```

Comment:
doesn't this attempt only for (maxAttempt-1) times

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/01b5e2f5f849d7a4e89d792e28845597544fca39